File: auto_verbs/verbs/ebay_com/ebay_search.py
"""
eBay – Vintage Mechanical Keyboard Search
Search: "vintage mechanical keyboard" | Filter: Buy It Now | Sort: Price + Shipping lowest

Pure Playwright – no AI. Uses .s-item CSS class selectors discovered via exploration.
"""

import re
import os
import shutil
import tempfile
import traceback
from playwright.sync_api import Playwright, sync_playwright

import sys as _sys
import os as _os
_sys.path.insert(0, _os.path.join(_os.path.dirname(__file__), ".."))
from cdp_utils import get_free_port, get_temp_profile_dir, launch_chrome, wait_for_cdp_ws

QUERY = "vintage mechanical keyboard"
MAX_RESULTS = 5
URL = "https://www.ebay.com/sch/i.html?_nkw=vintage%20mechanical%20keyboard&LH_BIN=1&_sop=15"


def get_temp_profile_dir(site="ebay"):
    """Create a temp Chrome profile dir to avoid locking the real one."""
    tmp = os.path.join(tempfile.gettempdir(), f"{site}_chrome_profile_{os.getpid()}")
    os.makedirs(tmp, exist_ok=True)
    src = os.path.join(
        os.environ.get("LOCALAPPDATA", ""),
        "Google", "Chrome", "User Data", "Default",
    )
    for f in ["Preferences", "Local State"]:
        s = os.path.join(src, f)
        if os.path.exists(s):
            shutil.copy2(s, os.path.join(tmp, f))
    return tmp


def dismiss_popups(page):
    """Dismiss cookie / GDPR popups."""
    for sel in [
        "#gdpr-banner-accept",
        "button:has-text('Accept')",
        "button:has-text('Accept All')",
        "button:has-text('Got it')",
        "button:has-text('OK')",
    ]:
        try:
            loc = page.locator(sel).first
            if loc.is_visible(timeout=800):
                loc.evaluate("el => el.click()")
                page.wait_for_timeout(300)
        except Exception:
            pass


def run(
    playwright: Playwright,
    search_query: str = QUERY,
    max_results: int = MAX_RESULTS,
) -> list:
    print("=" * 60)
    print("  eBay – Vintage Mechanical Keyboard Search")
    print("=" * 60)
    print(f'  Query: "{search_query}"')
    print(f"  Filter: Buy It Now | Sort: Price + Shipping lowest")
    print(f"  Max results: {max_results}\n")

    port = get_free_port()
    profile_dir = get_temp_profile_dir("ebay_com")
    chrome_proc = launch_chrome(profile_dir, port)
    ws_url = wait_for_cdp_ws(port)
    browser = playwright.chromium.connect_over_cdp(ws_url)
    context = browser.contexts[0]
    page = context.pages[0] if context.pages else context.new_page()
    results = []

    try:
        print("STEP 1: Navigate to eBay search results...")
        page.goto(URL, wait_until="domcontentloaded", timeout=30000)
        page.wait_for_timeout(2000)
        dismiss_popups(page)
        print(f"   Loaded: {page.url}\n")

        # Wait for results to render — try most likely selector first
        loaded = False
        for wait_sel in [".srp-results", "li.s-item", ".s-item__title", "[data-viewport]"]:
            try:
                page.wait_for_selector(wait_sel, timeout=5000)
                print(f"   ✅ Selector '{wait_sel}' appeared")
                loaded = True
                break
            except Exception:
                pass

        if not loaded:
            print("   ⚠ No known result selector appeared — will try fallbacks")

        # Scroll to trigger lazy loading
        for _ in range(2):
            page.evaluate("window.scrollBy(0, 600)")
            page.wait_for_timeout(500)
        page.evaluate("window.scrollTo(0, 0)")
        page.wait_for_timeout(500)

        print("STEP 2: Extract product listings...")

        # ──────────────────────────────────────────────────────
        # Strategy 1: Single JS evaluate — extracts all at once, no per-element timeouts
        # ──────────────────────────────────────────────────────
        skip_phrases = ["shop on ebay", "picks for you", "results matching", "related:", "save this search", "trending on", "see all", "sponsored"]

        all_items = page.evaluate("""(max) => {
            const results = [];
            const skip = ["shop on ebay", "picks for you", "results matching", "related:", "save this search", "trending on", "see all", "sponsored"];

            // Try .s-item first, then .srp-results > li
            let items = document.querySelectorAll('li.s-item');
            if (items.length === 0) items = document.querySelectorAll('.srp-results > li');

            for (const item of items) {
                if (results.length >= max) break;

                // Title: try .s-item__title, then role=heading, then first link
                let title = '';
                const titleEl = item.querySelector('.s-item__title') || item.querySelector('[role="heading"]');
                if (titleEl) title = titleEl.innerText.trim();
                if (!title) {
                    const linkEl = item.querySelector('a.s-item__link, a');
                    if (linkEl) title = linkEl.innerText.trim();
                }
                // Clean "Opens in a new window or tab"
                title = title.replace(/Opens in a new window or tab/gi, '').trim();

                if (!title || title.length < 5) continue;
                const lower = title.toLowerCase();
                if (skip.some(s => lower.startsWith(s))) continue;

                // Price
                let price = '';
                const priceEl = item.querySelector('.s-item__price') || item.querySelector('[class*="price"]');
                if (priceEl) price = priceEl.innerText.trim();
                if (!price) {
                    const m = item.innerText.match(/\\$(\\d[\\d,.]*)/);
                    if (m) price = '$' + m[1];
                }

                // Shipping
                let shipping = 'N/A';
                const shipEl = item.querySelector('.s-item__shipping') || item.querySelector('.s-item__freeXDays') || item.querySelector('[class*="shipping"]');
                if (shipEl) shipping = shipEl.innerText.trim();
                if (shipping === 'N/A') {
                    const lines = item.innerText.split('\\n');
                    for (const l of lines) {
                        if (/shipping|free\\s/i.test(l.trim())) { shipping = l.trim(); break; }
                    }
                }

                if (title && price) results.push({ title: title.substring(0, 120), price, shipping });
            }
            return results;
        }""", max_results)

        if all_items:
            results = all_items
            print(f"   ✅ JS evaluate extracted {len(results)} items")

        # ──────────────────────────────────────────────────────
        # Strategy 2: Fallback — full page text with regex
        # ──────────────────────────────────────────────────────
        if not results:
            print("   ⚠ JS evaluate returned 0 — trying text fallback...")
            body = page.locator("body").inner_text(timeout=15000)
            lines = [l.strip() for l in body.split("\n") if l.strip()]

            i = 0
            while i < len(lines) and len(results) < max_results:
                line = lines[i]
                if (len(line) >= 20 and not line.startswith("$")
                    and not any(line.lower().startswith(p) for p in skip_phrases)
                    and any(c.isalpha() for c in line)):
                    price = ""
                    shipping = "N/A"
                    for j in range(0, 6):
                        if i + j >= len(lines):
                            break
                        nxt = lines[i + j]
                        pm = re.search(r"\$[\d,.]+", nxt)
                        if pm and not price:
                            price = pm.group(0)
                        if re.search(r"(shipping|free\s)", nxt, re.IGNORECASE) and shipping == "N/A":
                            shipping = nxt.strip()[:80]
                    if price:
                        results.append({"title": line[:120], "price": price, "shipping": shipping})
                        i += 5
                    else:
                        i += 1
                else:
                    i += 1

        print(f"\n" + "=" * 60)
        print(f"  DONE – {len(results)} results")
        print("=" * 60)
        for i, r in enumerate(results, 1):
            print(f"  {i}. {r['title']}")
            print(f"     Price:    {r['price']}")
            print(f"     Shipping: {r['shipping']}")
            print()

    except Exception as e:
        print(f"\nError: {e}")
        traceback.print_exc()
    finally:
        try:
            browser.close()
        except Exception:
            pass
        chrome_proc.terminate()
        shutil.rmtree(profile_dir, ignore_errors=True)

    return results


if __name__ == "__main__":
    with sync_playwright() as playwright:
        items = run(playwright)
        print(f"Total results: {len(items)}")

File: auto_verbs/verbs-batch1/ebay_com/ebay_search.js
/**
 * eBay – Vintage Mechanical Keyboard Search
 *
 * Prompt:
 *   Search "vintage mechanical keyboard", filter "Buy It Now",
 *   sort "Price + Shipping: lowest first", top 5 (title, price, shipping).
 *
 * Strategy:
 *   Direct URL with params: ebay.com/sch/i.html?_nkw=vintage+mechanical+keyboard&LH_BIN=1&_sop=15
 *   _sop=15 = Price + Shipping: lowest first
 *   LH_BIN=1 = Buy It Now
 *   Then AI extract + record selectors for Python.
 */

const { Stagehand } = require("@browserbasehq/stagehand");
const { z } = require("zod");
const fs = require("fs");
const path = require("path");
const os = require("os");
const { setupLLMClient, PlaywrightRecorder, observeAndAct } = require("../../stagehand-utils");

/* ── kill switch ─────────────────────────────────────────── */
const TIMEOUT = 150_000;
setTimeout(() => { console.error("\n⏰ Global timeout – exiting"); process.exit(1); }, TIMEOUT);

/* ── config ──────────────────────────────────────────────── */
const CFG = {
  query:    "vintage mechanical keyboard",
  maxItems: 5,
  url() {
    return `https://www.ebay.com/sch/i.html?_nkw=${encodeURIComponent(this.query)}&LH_BIN=1&_sop=15`;
  },
};

/* ── temp Chrome profile ─────────────────────────────────── */
function getTempProfileDir(site = "ebay") {
  const tmp = path.join(os.tmpdir(), `${site}_chrome_profile_${Date.now()}`);
  fs.mkdirSync(tmp, { recursive: true });
  const src = path.join(os.homedir(), "AppData", "Local", "Google", "Chrome", "User Data", "Default");
  for (const f of ["Preferences", "Local State"]) {
    const s = path.join(src, f);
    if (fs.existsSync(s)) fs.copyFileSync(s, path.join(tmp, f));
  }
  return tmp;
}

/* ── genPython ───────────────────────────────────────────── */
function genPython(results) {
  const ts = new Date().toISOString();
  return `"""
eBay – Vintage Mechanical Keyboard Search
Search: "${CFG.query}" | Filter: Buy It Now | Sort: Price + Shipping lowest
Generated: ${ts}

Pure Playwright – no AI. Uses .s-item CSS class selectors discovered via exploration.
"""

import re
import os
import traceback
from playwright.sync_api import Playwright, sync_playwright

QUERY = "${CFG.query}"
MAX_RESULTS = ${CFG.maxItems}
URL = "https://www.ebay.com/sch/i.html?_nkw=${encodeURIComponent(CFG.query)}&LH_BIN=1&_sop=15"


def dismiss_popups(page):
    """Dismiss cookie / GDPR popups."""
    for sel in [
        "#gdpr-banner-accept",
        "button:has-text('Accept')",
        "button:has-text('Accept All')",
        "button:has-text('Got it')",
        "button:has-text('OK')",
    ]:
        try:
            loc = page.locator(sel).first
            if loc.is_visible(timeout=800):
                loc.evaluate("el => el.click()")
                page.wait_for_timeout(300)
        except Exception:
            pass


def run(
    playwright: Playwright,
    search_query: str = QUERY,
    max_results: int = MAX_RESULTS,
) -> list:
    print("=" * 60)
    print("  eBay – Vintage Mechanical Keyboard Search")
    print("=" * 60)
    print(f'  Query: "{search_query}"')
    print(f"  Filter: Buy It Now | Sort: Price + Shipping lowest")
    print(f"  Max results: {max_results}\\n")

    user_data_dir = os.path.join(
        os.environ["USERPROFILE"],
        "AppData", "Local", "Google", "Chrome", "User Data", "Default",
    )
    context = playwright.chromium.launch_persistent_context(
        user_data_dir,
        channel="chrome",
        headless=False,
        viewport={"width": 1280, "height": 900},
        args=[
            "--disable-blink-features=AutomationControlled",
            "--disable-infobars",
            "--disable-extensions",
        ],
    )
    page = context.pages[0] if context.pages else context.new_page()
    results = []

    try:
        print("STEP 1: Navigate to eBay search results...")
        page.goto(URL, wait_until="domcontentloaded", timeout=30000)
        page.wait_for_timeout(3000)
        dismiss_popups(page)
        print(f"   Loaded: {page.url}\\n")

        print("STEP 2: Extract product listings...")
        # eBay uses .s-item class for each listing card
        items = page.locator("li.s-item, div.s-item__wrapper").all()
        print(f"   Found {len(items)} .s-item elements")

        for item in items:
            if len(results) >= max_results:
                break
            try:
                # Title: .s-item__title
                title_el = item.locator(".s-item__title").first
                title = title_el.inner_text(timeout=2000).strip() if title_el.is_visible(timeout=500) else ""

                # Skip placeholder/header items
                if not title or title.lower().startswith("shop on ebay") or len(title) < 5:
                    continue

                # Price: .s-item__price
                price = ""
                try:
                    price_el = item.locator(".s-item__price").first
                    price = price_el.inner_text(timeout=1000).strip()
                except Exception:
                    pass

                # Shipping: .s-item__shipping or .s-item__freeXDays
                shipping = "N/A"
                try:
                    ship_el = item.locator(".s-item__shipping, .s-item__freeXDays").first
                    shipping = ship_el.inner_text(timeout=1000).strip()
                except Exception:
                    pass

                results.append({
                    "title": title,
                    "price": price,
                    "shipping": shipping,
                })
            except Exception:
                continue

        print(f"\\n" + "=" * 60)
        print(f"  DONE – {len(results)} results")
        print("=" * 60)
        for i, r in enumerate(results, 1):
            print(f"  {i}. {r['title']}")
            print(f"     Price:    {r['price']}")
            print(f"     Shipping: {r['shipping']}")
            print()

    except Exception as e:
        print(f"\\nError: {e}")
        traceback.print_exc()
    finally:
        context.close()

    return results


if __name__ == "__main__":
    with sync_playwright() as playwright:
        items = run(playwright)
        print(f"Total results: {len(items)}")
`;
}

/* ── popup dismissal ─────────────────────────────────────── */
async function dismissPopups(page) {
  const sels = [
    "#gdpr-banner-accept",
    "button:has-text('Accept')",
    "button:has-text('Accept All')",
    "button:has-text('Got it')",
    "button:has-text('OK')",
  ];
  for (const s of sels) {
    try {
      const el = page.locator(s).first();
      if (await el.isVisible({ timeout: 600 })) {
        await el.click({ timeout: 1_000 });
        await page.waitForTimeout(300);
      }
    } catch {}
  }
}

/* ── search (direct URL) ────────────────────────────────── */
async function searchEbay(stagehand, page, recorder) {
  console.log(`\n🔍 Searching eBay for "${CFG.query}"...`);
  console.log(`   URL: ${CFG.url()}`);

  await page.goto(CFG.url(), { waitUntil: "domcontentloaded", timeout: 30_000 });
  recorder.record("goto", `Navigate to ${CFG.url()}`);

  await page.waitForTimeout(3_000);
  recorder.record("wait", "Wait for eBay search results");

  console.log(`   ✅ Results loaded: ${page.url()}`);
}

/* ── extract ─────────────────────────────────────────────── */
async function extractResults(stagehand, page, recorder) {
  console.log("🎯 Extracting top 5 results...\n");

  const schema = z.object({
    items: z.array(z.object({
      title:    z.string().describe("Product title"),
      price:    z.string().describe("Price with $ sign"),
      shipping: z.string().describe("Shipping cost or 'Free shipping'"),
    })).describe(`Top ${CFG.maxItems} vintage mechanical keyboards`),
  });

  const MAX_TRIES = 3;
  for (let t = 1; t <= MAX_TRIES; t++) {
    console.log(`   Attempt ${t}: Extracting...`);
    try {
      const { items } = await stagehand.extract(
        `Extract the top ${CFG.maxItems} product listings. For each, get the title, price (with $ sign), and shipping cost. Skip any "Shop on eBay" header items.`,
        schema,
      );
      if (items && items.length > 0) {
        console.log(`   ✅ Extracted ${items.length} results on attempt ${t}`);
        recorder.record("extract", `Extract top ${CFG.maxItems} items (title, price, shipping)`);
        return items;
      }
    } catch (e) {
      console.log(`   ⚠ Attempt ${t} error: ${e.message}`);
    }
    if (t < MAX_TRIES) {
      await page.evaluate(() => window.scrollBy(0, 600));
      await page.waitForTimeout(2_000);
    }
  }

  console.log("   ❌ Extraction failed after retries");
  return [];
}

/* ── main ────────────────────────────────────────────────── */
(async () => {
  console.log("═══════════════════════════════════════════════════════════════");
  console.log("  eBay – Vintage Mechanical Keyboard Search");
  console.log("═══════════════════════════════════════════════════════════════");
  console.log(`  ⌨️  Search: "${CFG.query}"`);
  console.log(`  📊 Sort: Price + Shipping lowest | Filter: Buy It Now`);
  console.log(`  📦 Extract up to ${CFG.maxItems} results\n`);

  console.log("🤖 Setting up GitHub Models API...");
  const llmClient = setupLLMClient("hybrid");
  console.log("✅ GitHub Models API ready\n");

  console.log("🎭 Initializing Stagehand...");
  const tmpProfile = getTempProfileDir();
  console.log(`📁 Temp profile: ${tmpProfile}`);

  const stagehand = new Stagehand({
    env: "LOCAL",
    verbose: 0,
    llmClient,
    localBrowserLaunchOptions: {
      headless: false,
      args: [
        `--user-data-dir=${tmpProfile}`,
        "--disable-blink-features=AutomationControlled",
      ],
    },
  });
  await stagehand.init();
  console.log("✅ Stagehand ready\n");

  const page = stagehand.context.pages()[0];
  const recorder = new PlaywrightRecorder();

  try {
    await searchEbay(stagehand, page, recorder);

    console.log("🔲 Dismissing popups...");
    await dismissPopups(page);

    await page.evaluate(() => window.scrollBy(0, 300));
    await page.waitForTimeout(1_500);

    const results = await extractResults(stagehand, page, recorder);

    console.log(`📋 Found ${results.length} results:`);
    results.forEach((r, i) => {
      console.log(`   ${i + 1}. ${r.title}`);
      console.log(`      Price:    ${r.price}`);
      console.log(`      Shipping: ${r.shipping}\n`);
    });

    console.log("═══════════════════════════════════════════════════════════");
    console.log(`  ✅ DONE — ${results.length} results`);
    console.log("═══════════════════════════════════════════════════════════");
    results.forEach((r, i) => {
      console.log(`  ${i + 1}. ${r.title}`);
      console.log(`     Price:    ${r.price}`);
      console.log(`     Shipping: ${r.shipping}`);
    });

    // genPython disabled — ebay_search.py is hand-optimized (page.evaluate + temp profile)
    // const pyPath = path.join(__dirname, "ebay_search.py");
    // fs.writeFileSync(pyPath, genPython(results), "utf-8");
    // console.log(`\n✅ Python saved: ${pyPath}`);

    const actPath = path.join(__dirname, "recorded_actions.json");
    fs.writeFileSync(actPath, JSON.stringify(recorder.actions, null, 2), "utf-8");
    console.log(`📋 Actions saved: ${actPath}`);

  } finally {
    console.log("\n🧹 Closing...");
    await stagehand.close();
    fs.rmSync(tmpProfile, { recursive: true, force: true });
    console.log("🎊 Done!");
  }
})();
